### DISCUSSION TITLE
AllValuesMultipleFilter raises 404 error if any option does not exist

### DISCUSSION THREAD

**Discussion by palomapiot** [carltongibson/django-filter#1420] [Q&A]
I have a search feature with on multiple fields (age, gender, location, etc.) and I'm using `AllValuesMultipleFilter` with `lookup_expr` `contains` to do the searches. The options among each field are acting like an OR, but among different fields, it is an AND, as expected.

My current problem is when I trigger a search and I pick a value (ex. age 12-20) which is NOT in any entity, instead of ignoring it and returning either nothing or the other matches (ex. if I also choose gender female, all entities where gender is female) I am getting a 404 error and my app is going to the 404 page.

Is there any way to solve this? Or should this be reported as a bug?

Profile filter set:
```
class ProfileFilterSet(filters.FilterSet):
    validated_data__age = filters.AllValuesMultipleFilter(
        field_name='validated_data__age',
        lookup_expr='contains'
    )
    validated_data__gender = filters.AllValuesMultipleFilter(
        field_name='validated_data__gender',
        lookup_expr='contains'
    )
    validated_data__location = filters.AllValuesMultipleFilter(
        field_name='validated_data__location',
        lookup_expr='contains'
    )

    class Meta:
        model = Profile
        fields = ['validated_data__age', 'validated_data__gender', 'validated_data__location', 'experiment_id']
```

Serializer:
```
class ProfileViewSet(viewsets.ModelViewSet):
    queryset = Profile.objects.all().order_by('experiment_id')
    serializer_class = ProfileSerializer
    permission_classes = [permissions.IsAuthenticated]
    filter_backends = [DjangoFilterBackend]

    filter_class = ProfileFilterSet
```

Model:
```
class Profile(models.Model):
    experiment_id = models.TextField(max_length=50)
    validated_data = models.ForeignKey(
        ProfileData, 
        on_delete=models.CASCADE,
        related_name='%(class)s_validated_data',
        blank=True,
        null=True
    )
```

```
class ProfileData(models.Model):

    class Age(models.TextChoices):
        ADOLESCENCE = '12-20', _('12-20')
        EARLY_ADULTHOOD = '20-30', _('20-30')
        MATURE_ADULTHOOD = '30-65', _('30-65')
        OLD_AGE = 'Over65', _('Over65')
        UNKNOWN = 'Unknown', _('Unknown')

    class Gender(models.TextChoices):
        MALE = 'Male', _('Male')
        FEMALE = 'Female', _('Female')
        UNKNOWN = 'Unknown', _('Unknown')

    age = models.TextField(
        choices=Age.choices,
        default=Age.UNKNOWN
    )
    gender = models.TextField(
        choices=Gender.choices,
        default=Gender.UNKNOWN
    )
    location = CountryField(blank_label='', blank=True, null=True)

    def __str__(self):
        """String for representing the Model object."""
        return '%s, %s, %s' % (self.age, self.gender, self.location)
```

Thank you.

**Comment by carltongibson** [comment]
Sounds as if you’re filtering to an empty QuerySet, and so DRF is (as expected) raising a 404 error? 🤔

**Comment by palomapiot** [comment]
The problem is, for example, if I select the filter age `12-20` and `20-30` and I don't have any users in the `12-20` age group, I get back a 404 Not Found Error. But I think the expected behaviour should be returning the users with age `20-30` **or** an empty list if there aren't any matches, but not raising the 404. Because then in my app, I get the redirection to the 404 page instead of having an empty list to show.

Maybe there is a way to solve this, but, unfortunately, I wasn't able to find it. Thank you so much for your reply.

**Reply by carltongibson** [reply]
Can you dig in and see exactly where the 404 is being raised? (I suspect it's in the DRF view... — that's OK, but it's hard to say what to adjust without the traceback. ) Thanks

**Comment by normic** [comment]
I stumbled about this today and had the same thought (404 or empty list). I think the 404 is correct here, as we are referring to data which is not in the pre-filtered list.

**Comment by normic** [comment]
In my case, the frontend is using data from an endpoint with all active user(ids).
But when selecting a user (for filtering) which has no entry in the data to be filtered, one get's the mentioned 404.

A look into the browsable API shows that we have a correct list with all possible and valid values.
In my opinion the correct way would be to use this pre-filtered list in the frontend as data for the filterable values.

Question: Is there a way to get the pre-filtered data into the view? Then the frontend could simply use this. I guess it's possible to do some magic with accessing the filterset_class, but is there a recommend way?

**Reply by carltongibson** [reply]
@normic I've read this several times and don't really understand exactly what you're asking. I'd suggest you start a fresh discussion and explain it afresh, rather than picking up a three year old conversation.

**Reply by normic** [reply]
Hi @carltongibson thanks for your reply. 
I've just tried to put everything together in a hopefully more understandable way. 
New thread is #1692
